docs/implementing-real-time-reporting.md

```diff
@@ -0,0 +1,102 @@
+
+Once this report has been generated, it is recommended to automatically open this report for a better user experience.
+
+_<sup>1</sup> This could change in the future. Instead, the only thing that would be necessary is to show the files without any mutants and add new mutants while they are being generated. This would allow the report to open earlier._
+
+### Changes to the `<mutation-test-report-app>` component
+
+To inform the report that it is going to be provided with updates, set the new `sse` property with a URL to where the SSE endpoint is located.
+
+```html
+<mutation-test-report-app sse="localhost:8080/sse"></mutation-test-report-app>
+```
+
+### Server side implementation
+
+At the server side, several HTTP headers are necessary to let the browser know that it should keep the connection open:
+
+```http
+Content-Type: text/event-stream
+Cache-Control: no-cache
+Connection: keep-alive
+Access-Control-Allow-Origin: *
+```
+
+The `Access-Control-Allow-Origin: *` is necessary only if the report is opened from a file from disk. It is possible to be more strict with CORS if the report is hosted.
+
+When sending events to the browser, please note that the payload should look like the following:
+
+```
+event: eventName\n
+data: "..."\n\n
+```
+
+For more information, refer to the [documentation](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format).
+
+### Events
+
+The report currently listens to two events:
+
+- `mutant-tested`.
+- `finished`.
+
+#### The `mutant-tested` event
+
+This event should be sent when a single mutant has been tested. The mutant should conform to the [report schema](https://github.com/stryker-mutator/mutation-testing-elements/blob/master/packages/report-schema/src/mutation-testing-report-schema.json).
+
+Required properties:
+
+- `id`.
+- `status`.
+
+Optional properties:
+
+- `coveredBy`.
+- `description`.
+- `duration`.
+- `killedBy`.
+- `location`.
+- `mutatorName`.
+- `replacement`.
+- `static`.
+- `statusReason`.
+- `testsCompleted`.
+
+For this event the `data` field in the event should be stringified JSON:
+
+```http
+event: mutant-tested
+data: "{\"id\":\"0\",\"status\":\"Killed\",\"coveredBy\":[\"test_1\"]}"
+```
+
+#### The `finished` event
+
+To notify that the report should stop listening for events, send this event:
+
+```http
+event: finished
+data: null
+```
+
+## Limitations
+
+Please note that if a user refreshes the page while realtime reporting is ongoing, all the state will be lost. To keep this state, the following can be done:
+
+1. Keep track of every mutant event that has been tested. When a browser reconnects, send all previous mutant results.
+2. If the report is opened from a file, it is possible to write the report to disk once the mutation testing process is done. If the user refreshes the page, they will see the full report.
```

Comment:
We thought about that but then you'd need to somehow redirect them to the file when the testrun is finished, or keep the webserver open.

Reply:
The main issue with hosting the report is what you do when the reporting is done. Personally I don't think redirecting the user to a file is nice UX.

Reply:
If we want we can have a discussion about UX after the report is done at a later time. For now, I would keep this out of the docs (as stated in [the discussion above](https://github.com/stryker-mutator/mutation-testing-elements/pull/2505#discussion_r1189508824))

Reply:
> Whenever the report is opened, first push the 'base' state to SSE

I don't think this will work, as I talked about [here](https://github.com/stryker-mutator/mutation-testing-elements/pull/2505#discussion_r1189029301). 

But it could work in combination with the report property. It is possible to keep track of mutants that have been tested and serve those when a browser connects to the SSE endpoint (but it would require #2498 to be merged to be able to handle large amounts of events)

Reply:
Let's mark filling the initial state using events instead of the report property as a future improvement. It's more or less the same feature as discussed here: https://github.com/stryker-mutator/mutation-testing-elements/pull/2505/files/1c96cae494a30ef9d1dfb1189901a2e56cb4f4c0#r1189030710